File: ai-solutions/02/gemini-3-pro/python/main.py
import os

def solve():
    input_path = os.path.join(os.path.dirname(__file__), "../../../../inputs/02.txt")
    with open(input_path, 'r') as f:
        data = f.read().strip()

    ranges = []
    parts = data.split(',')
    for part in parts:
        start, end = map(int, part.split('-'))
        ranges.append((start, end))

    total_p1 = 0
    total_p2 = 0

    for start, end in ranges:
        for n in range(start, end + 1):
            s = str(n)
            length = len(s)
            
            # Part 1 Check
            is_p1 = False
            if length % 2 == 0:
                half = length // 2
                if s[:half] == s[half:]:
                    is_p1 = True
                    total_p1 += n
            
            # Part 2 Check
            # Check if composed of repeated pattern of length L
            # L must verify: length % L == 0 and length // L >= 2
            is_p2 = False
            for L in range(1, length // 2 + 1):
                if length % L == 0:
                    pattern = s[:L]
                    times = length // L
                    if pattern * times == s:
                        is_p2 = True
                        total_p2 += n
                        break
    
    print(total_p1)
    print(total_p2)

if __name__ == "__main__":
    solve()

File: ai-solutions/02/gemini-3-pro/rust/src/main.rs
use std::fs;
use std::path::Path;

fn main() {
    let input_path = Path::new("../../../../inputs/02.txt");
    let content = fs::read_to_string(input_path).expect("Failed to read input file");
    let content = content.trim();

    let mut ranges = Vec::new();
    for part in content.split(',') {
        if let Some((start_str, end_str)) = part.split_once('-') {
            let start: u64 = start_str.parse().expect("Invalid start number");
            let end: u64 = end_str.parse().expect("Invalid end number");
            ranges.push((start, end));
        }
    }

    let mut total_p1: u64 = 0;
    let mut total_p2: u64 = 0;

    for (start, end) in ranges {
        for n in start..=end {
            let s = n.to_string();
            let len = s.len();

            // Part 1
            if len % 2 == 0 {
                let half = len / 2;
                let (first, second) = s.split_at(half);
                if first == second {
                    total_p1 += n;
                }
            }

            // Part 2
            for l in 1..=(len / 2) {
                if len % l == 0 {
                    let pattern = &s[0..l];
                    let times = len / l;
                    if pattern.repeat(times) == s {
                        total_p2 += n;
                        break;
                    }
                }
            }
        }
    }

    println!("{}", total_p1);
    println!("{}", total_p2);
}
